### PR TITLE
expose vulkan physical device capabilities, enabled device extensions

### PR DESCRIPTION
**Pull request by i509VCB**
**Description**
Well this pull request expanded a bit as you can see in the edits.

- Expose physical device properties and PhysicalDeviceCapabilities from Vulkan's Adapter
- For Vulkan's Device, expose the following
  - Enabled device extensions
  - Raw physical device (needed for getting data associated with a device that is too varied to calculate at creation time).
  - The instance the device is from (for getting said physical device data)
  - Version of the instance on the device (For above)
  - Enabled instance extensions (Also for above)
- For Vulkan's Adapter, expose the following
  - The instance
  - The version of the instance
  - Enabled instance extensions

### CODE REVIEW

File: wgpu-hal/src/vulkan/device.rs [hunk 1]
```diff
@@ -699,6 +697,27 @@ impl super::Device {
...
+    pub fn raw_instance(&self) -> &ash::Instance {
+        &self.shared.instance.raw
+    }
+
+    /// Returns the driver api version of the instance.
+    pub fn driver_api_version(&self) -> u32 {
+        self.shared.instance.driver_api_version
+    }
+
+    pub fn enabled_instance_extensions(&self) -> &[&'static CStr] {
```
Comment (cwfitzgerald) [inline wgpu-hal/src/vulkan/device.rs:718]: This feels weird, considering you can just call this function on the instance.

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
Baring one comment, I think this is fine. My only question is more design focused, should we be exposing things like this or should we be making the fields public, which would potentially allow creating wgpu-hal types from the raw vulkan stuff.

File: wgpu-hal/src/vulkan/device.rs (hunk 1 above)
Comment (i509VCB) [reply wgpu-hal/src/vulkan/device.rs:718]: Yes that could be done. However the wgpu apis implicitly hide the instance when the device is created.

One option could be to make the underlying InstanceShared accessible so we don't need to duplicate so many functions for instance fields.

File: wgpu-hal/src/vulkan/mod.rs [hunk 2]
```diff
@@ -79,8 +79,9 @@ struct DebugUtils {
     messenger: vk::DebugUtilsMessengerEXT,
 }
 
-struct InstanceShared {
+pub struct InstanceShared {
```
[the remark this answers is not in the document]
Comment (i509VCB) [reply wgpu-hal/src/vulkan/mod.rs:82]: I added some functions to InstanceShared to access the fields like entry, instance, extensions, driver api version.

Plus this has to be public if we declare this as a return type in `Instance` or else rustc won't let it slide.

[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply wgpu-hal/src/vulkan/mod.rs:82]: I missed your other changes, this is fine

**Review by cwfitzgerald — APPROVED** [review]
TYSM

**Pull request merged**